docs/docs/custom-graph-components.mdx

```diff
@@ -0,0 +1,432 @@
+      Rasa Open Source will use the graph component's `train` method if
+      `is_trainable=True` is specified. Rasa Open Source will use
+      the graph component's `predict_action_probabilities` to make predictions for the
+      next action which should be run within a conversation.
+
+- `is_trainable`: Specifies if the graph component requires to train itself before it can
+   process training data for other dependent graph components or before it can make
+   predictions
+
+- `model_from`: Specifies if a pretrained
+   [language model](components.mdx#language-models) needs to be provided to the `train`,
+  `process_training_data` and `process` methods of the graph component. These methods
+   have to support the parameter `model` to receive the language model. Note that you
+   still need to make sure that the model provided graph component is part of your model
+   configuration.
+
+## Using Custom Components in your Model Configuration
+You can use custom components like any other NLU component or policy within your
+[model configuration](model-configuration.mdx). The only change is that you have to specify
+the full module name instead of the class name only. The full module name depends on
+your module's location in relation to the specified
+[PYTHONPATH](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH).
+By default, Rasa Open Source adds the directory from where you run the CLI to the
+`PYTHONPATH`. If you e.g. run the CLI from `/Users/<user>/my-rasa-project`
+and your module `MyComponent` is in `/Users/<user>/my-rasa-project/custom_components/my_component.py`
+then the module path is `custom_components.my_component.MyComponent`. Everything except
+the `name` entry will be passed as `config` to your component.
+
+```yaml-rasa {5-6,10} title="config.yml"
+recipe: default.v1
+language: en
+pipeline:
+# other NLU components
+- name: your.custom.NLUComponent
+  dimensions: ["number"]
+
+policies:
+# other dialogue policies
+- name: your.custom.Policy
+```
+
+## Implementation Hints
+
+### Message Metadata
+When you [define metadata for your intent examples in your training data](./training-data-format.mdx#training-examples),
+your NLU component can access both the intent metadata and the intent example metadata during processing:
+
+```python
+# in your component class
+
+    def process(self, message: Message, **kwargs: Any) -> None:
+        metadata = message.get("metadata")
+        print(metadata.get("intent"))
+        print(metadata.get("example"))
+```
+
+### Sparse and Dense Message Features
+If you create a custom message featurizer you can return two different kind of features: sequence features and sentence
+features. The sequence features are a matrix of size `(number-of-tokens x feature-dimension)`, e.g.
+the matrix contains a feature vector for every token in the sequence.
+The sentence features are represented by a matrix of size `(1 x feature-dimension)`.
+
+## Examples of Custom Components
+Vincent?
```

Comment:
> If I could pick only two, I might add the Dense Featurizer and the Trainable Sparse Featurizer.

Then let's do these and add links to the others?

Reply:
Sure thing! I'm at a conference now but I'll review the components one more time and make a PR. 

Note: my plan is now also to make a course (instead of a blogpost) on the learning center where I'll showcase the rest.